### ISSUE TITLE
{($checked)} binding on named radio groups doesn't update keys for deselected options.

### ISSUE THREAD

**Issue by bmomberger-bitovi** [canjs/can-stache-bindings#125]
When changing to another option in a named radio group (any set of radio buttons in a form that have the same value for the `name` attribute), the `change` event only fires on the newly selected radio input.

For elements cross-bound to scope properties via the element's `checked` property (i.e. the basic case for input cross-binding, not using any converters), this means that once selected, a boolean property cross-linked to an element is always true, even after the element being deselected.

http://jsbin.com/nayoyabira/1/edit?html,js,output for reference.

Proposed solution:  when binding `$checked`, look for the presence of a `name` attribute on the element, and if it exists, modify the handler to update the state of *all* scope properties tied to same-named radio buttons.

**Comment by justinbmeyer** [comment]
I'm thinking we might need to use some sort of event delegation to make this work, likely with a custom event binding.  

For example, we could allow people to `radioElement.addEventListener("groupchange", handler)`. This would setup "change" event delegation on the documentElement if it hasn't already been established.  This delegation would see if a radio element has changed, and then call any registered `handlers` on associated radio elements.

**Comment by matthewp** [comment]
I don't think that's the best way to deal with this problem. can-stache-bindings used to have a lot of conditional logic but that was removed in 3.0 and now only has one special case for backwards compat IIRC.

I think adding special casing of events and attributes to the template layer is confusing. It makes it hard to recognize the boundaries between the DOM APIs and the template's APIs. To the user it *looks* like they are using the DOM, but they are actually using something else.

I would instead:

* Throw on the use of two-way binding syntax for radio elements as it is not supported by the browser. This would solve the immediate issue of things not working as expected.
* Create a can-radio can-component that has the types of events that are desired. Since this is a new abstraction there's no confusion about how it works and enables us to make it be whatever we want it to be in the future.

**Comment by justinbmeyer** [comment]
`can-stache-bindings` is already attempting to solve the problem of “how to listen to a given attribute/property changing”.  We do this for `<input value/>` -> `change` events for example.  Imo, we should continue to do this for as many attributes/properties as make sense.  It's one of the shorthands our two-way bindings provide. 

But, we should document the `element/attribute||property->event` relationship.  Throwing for this one case isn't going solve other examples of this (like listening on paused on a `video` element).


I think the solution to your valid concern would be to have a different syntax that allows the user to control what event to bind to for a given property/attribute.  For example:

```html
<!-- listen to `oninput` events for when an input's value change -->
<input {value}(input)="key"/>

<!-- the current behavior -->
<input {value}(change)="key"/>

<!-- make radiochange work -->
<input type="radio" name="something" {checked}(radiochange)="key"/>

<!-- make paused property work -->
<input type="radio" name="something" {paused}(pause)="key"/>
```

_I'm very open to other suggested syntaxes_.


This might allow these conventional mappings to disappear over time, force users to know which events they are binding to if they don't explicitly match the name of the property.

**Comment by matthewp** [comment]
I don't think we should make such big changes to can-stache-bindings such as a new syntax given that we are likely going to replace can-stache with a different template syntax in the mid-term future. Adding new features puts a burden on a potentially better new template syntax as people will expect the same new features to be present in that engine.  I will be arguing against any sort of can-util/attr/attr types of abstractions from existing in the new template engine.

**Comment by justinbmeyer** [comment]
@matthewp If there's going to be two-way binding in a future template, something like this would arguably be better, allow more control in the current one.  It's worth adding even to learn from. 
 Plus, adding features for stache, when we have no agreed-upon or in-progress replacement, is totally fine (and painless (~10 LOC) b/c of the existing code-base). We should fix problems for users today and tomorrow. 

> I will be arguing against any sort of can-util/attr/attr types of abstractions from existing in the new template engine.

Does such a thing currently exist?  Almost certainly not in any two-way bound engine.  Even in one-way, I'd expect that something like react doesn't simply just set properties blindly.  It likely corrects for browser quirks, and things like changing `.value` might not actually select the right `<option>`.  There's probably some version of `.attr` that gets invoked when a change is detected and the state of the DOM needs to change.

**Comment by matthewp** [comment]
I'm ok with adding this as long as it is taken out of can-util and becomes a can-stache-bindings only feature, and there is no expectation that all can-stache/can-stache-bindings features (and use-cases) will survive into the next engine.  I will be arguing against the attr abstraction, like I said before.

I don't know how other frameworks handle this exact scenario. I know that many run into the same problem when they try to generalize something that is not general (form elements) and regret it.  Regardless, components are a well-understood abstraction for creating custom element types, including extending built-in types, so that feels like the better solution for making form elements that speak the canjs language.

**Comment by justinbmeyer** [comment]
@matthewp why out of `can-util`?  We are also adding an `enter` event in `can-util`.  These seem very similar, only differentiated by how likely someone is to use `enter` vs `radiochange`.

I guess I'm ok putting it in its own package.  How should we draw the line?  What about `inserted` and `removed`?

> can-stache/can-stache-bindings features (and use-cases) will survive into the next engine

I think there's some core stuff that needs to survive around basic translation between common dom elements (including form elements) and the data layer.  People know the DOM and would expect any framework to work with it (instead of having to learn the behaviors of other elements).

I agree that the closer to the metal you can get, the better.  But that's why I think something like: `(){}` SHOULD be in whatever template engine we provide if it performs two-way bindings.  It's exposing what's going on.  I'd go even further and add some way of distinguishing between properties and attributes.  `(){}` is generalizing a two-way binding behavior.

**Comment by justinbmeyer** [comment]
@matthewp I know that `focused` was an issue for react for a while.  How would you solve this w/o `attr`? `attr` has been a great for CanJS's code quality.  It's hard to understand why you'd want to remove it.

**Comment by matthewp** [comment]
@justinbmeyer and I talked about it and we agreed that moving custom events into their own packages in the future is the way to go.  For this event it will be moved to `can-event-radiochange`. can-stache-bindings will use that package instead of the implementation in can-util.

In the future we might migrate away from binding syntax including special events and instead have users bring those in.  I will argue for this in the next templating engine.

**Comment by matthewp** [comment]
Associated can-util issue: https://github.com/canjs/can-util/issues/245

**Comment by andrejewski** [comment]
Closed by #187 🔨
